### PR TITLE
Issue/3030

### PR DESCRIPTION
**Pull request by Alisu**
introduce 
 - ClassParser
 - ClassDefinition
 - UndefinedClasses

Integrate them with ClassBuilder (optional and off by default)
Update SugsSuggestions with new nodes

All with tests

### CODE REVIEW

**Comment by MarcusDenker** [comment]
we really need to carefully review this PR. the last had really odd side effects. e.g. it made class builder tests fail.

**Comment by guillep** [comment]
Hi Marcus, I'm taking care of it, please do not stress :)

**Comment by guillep** [comment]
The build now finishes ok, except for some failing tests. These remaining tests are

image

Are they some flaky tests?

**Comment by MarcusDenker** [comment]
The announcer test fails all couple of build. no idea why (see #2471) The testMaxLengthIsSetInWidget falls consistently (not related to the changes in this issue). I added https://github.com/pharo-project/pharo/issues/3119 for that.

File: src/ClassParser/ShiftClassBuilder.extension.st [hunk 1]
```diff
@@ -0,0 +1,35 @@
+Extension { #name : #ShiftClassBuilder }
+
+{ #category : #'*ClassParser' }
+ShiftClassBuilder >> buildFromAST: aCDClassDefinitionNode [ 
+ 
+|package|
```
Comment (guillep) [inline src/ClassParser/ShiftClassBuilder.extension.st]: Could you format this method ? It really does not follow any convention. See below

File: src/ClassParser/ShiftClassBuilder.extension.st [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	self name: aCDClassDefinitionNode className.
+	aCDClassDefinitionNode superclassName = #nil ifFalse:[
+	self superclass: (superclassResolver resolve: aCDClassDefinitionNode superclassName inEnv: buildEnvironment).]
+	ifTrue:[self superclass: nil].
```
Comment (guillep) [inline src/ClassParser/ShiftClassBuilder.extension.st]: Here again

File: src/ClassParser/ShiftClassBuilder.extension.st [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	aCDClassDefinitionNode isEphemeron ifTrue:[self layoutClass: EphemeronLayout].
+	aCDClassDefinitionNode isWeak ifTrue:[self layoutClass: WeakLayout].
+	aCDClassDefinitionNode isImmediate ifTrue:[self layoutClass: ImmediateLayout].
+	aCDClassDefinitionNode isVariable ifTrue:[self layoutClass: VariableLayout].
+	aCDClassDefinitionNode isWords ifTrue:[self layoutClass: WordLayout].
+	aCDClassDefinitionNode isBytes ifTrue:[self layoutClass: ByteLayout].
+	
+	self slots: (aCDClassDefinitionNode slots collect: [:e| e asSlot]).
+	self sharedVariables:(aCDClassDefinitionNode sharedSlots collect: [:e| e]).
+	aCDClassDefinitionNode packageName ifNotNil:[:cat | package:= cat] ifNil:[package:= 'Unclassified'].
```
Comment (guillep) [inline src/ClassParser/ShiftClassBuilder.extension.st]: Here again

File: src/ClassParser/ShiftClassBuilder.extension.st [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	self category: package.
+	self layoutDefinition sharedPools: (aCDClassDefinitionNode sharedPools collect: [:e | e name]).
+	
+	aCDClassDefinitionNode traitDefinition ifNotNil:[:traitDef| traitDef originalNode formattedCode.
+	self traitComposition: 
+	(self class compiler 
+	source: traitDef originalNode formattedCode;
+	"environment: buildEnvironment;"
+	evaluate)].
```
Comment (guillep) [inline src/ClassParser/ShiftClassBuilder.extension.st]: What is that line commented?

**Review by guillep — CHANGES_REQUESTED** [review]

**Comment by guillep** [comment]
Thanks Theo!

**Pull request merged**